src/dataflow/src/render/mod.rs

```diff
@@ -740,6 +740,70 @@ impl<G> Context<G, RelationExpr, Row, Timestamp>
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    fn extract_map_filter_project(
+        &mut self,
+        relation_expr: &RelationExpr,
+    ) -> Option<(MapFilterProject, RelationExpr)> {
+        match relation_expr {
+            RelationExpr::Map { input, scalars } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.map(scalars.iter().cloned()), input))
+            }
+            RelationExpr::Filter { input, predicates } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.filter(predicates.iter().cloned()), input))
+            }
+            RelationExpr::Project { input, outputs } => {
+                let (mfp, input) = self.extract_map_filter_project(input)?;
+                Some((mfp.project(outputs.iter().cloned()), input))
+            }
+            RelationExpr::Get { .. } => Some((
+                MapFilterProject::new(relation_expr.arity()),
+                relation_expr.clone(),
+            )),
+            _ => None,
+        }
+    }
+
+    /// Attempt to extract a chain of map/filter/project operators on top of a Get. Returns true if
+    /// it was successful and false otherwise. We still fall back to individual implementations for
+    /// the various operators so that ones like Filter, that special-case being on top of a join,
+    /// can still do so.
+    fn try_render_map_filter_project(
+        &mut self,
+        relation_expr: &RelationExpr,
+        scope: &mut G,
+        worker_index: usize,
+    ) -> bool {
+        if let Some((mfp, input)) = self.extract_map_filter_project(relation_expr) {
+            self.ensure_rendered(&input, scope, worker_index);
+            let (ok_collection, mut err_collection) = self
+                .flat_map_ref(&input, {
+                    let mut row_packer = repr::RowPacker::new();
+                    let temp_storage = RowArena::new();
+                    move |row| {
+                        mfp.evaluate(&mut row.unpack(), &temp_storage, &mut row_packer)
```

Comment:
Very minor, and not actionable, but this is the other moment at which a potentially optional allocation happens. `row.unpack()` will produce a `Vec<Datum>` that we could try and wrangle in to re-using an existing allocation. It's a bit awkward with lifetimes, but inside `flat_map_ref()` we could mess around a bit there and re-use one vector for each operator invocation.

Reply:
Just read the code, and it would at least involve copy/paste of differential's `flat_map_ref`, as that method does not expose the `'storage` lifetime, which is what we would want to use for the lifetime of the `Vec<Datum>`.